chatgpt_streaming.exs

```diff
@@ -0,0 +1,90 @@
+# Usage: elixir chatgpt_streaming.exs openai-api-key "prompt"
+
+Mix.install([
+  {:eventsource_ex, "~> 2.0"},
+  {:httpoison, "~> 1.5"},
+  {:jason, "~> 1.4"}
+])
+
+defmodule ChatGPTStreamer do
+  use GenServer
+
+  def start_link(state) do
+    GenServer.start_link(__MODULE__, state)
+  end
+
+  def init(state) do
+    {:ok, request_completion(state[:api_key], state[:messages], state[:parent_pid])}
+  end
+
+  def handle_info(
+        %EventsourceEx.Message{
+          data: "[DONE]",
+          event: "message"
+        },
+        parent_pid
+      ) do
+    IO.puts("")
+    send(parent_pid, :done)
+    {:stop, :normal, nil}
+  end
+
+  def handle_info(
+        %EventsourceEx.Message{
+          data: jason_payload,
+          event: "message"
+        },
+        state
+      ) do
+    message = Jason.decode!(jason_payload)
+    extract_text(message) |> IO.write()
+    {:noreply, state}
+  end
+
+  def extract_text(%{"choices" => choices}) do
+    choices
+    |> Enum.map(fn %{"delta" => delta} -> delta end)
+    |> Enum.filter(fn content -> Map.has_key?(content, "content") end)
+    |> Enum.map(fn %{"content" => content} -> content end)
+    |> Enum.join()
```

Comment:
I believe this is equivalent to the following which would do everything in a single pass:

```suggestion
    for %{"delta" => delta} <- choices,
        %{"content" => content} <- delta,
        into: "" do
      content
    end
```

could you double-check?

Reply:
Thanks @wojtekmach! That works with the addition of a filter expression:

https://github.com/wojtekmach/mix_install_examples/pull/22/commits/b937d00475116dd4369a34c1a668e15eefd593da#diff-42fd847355cc16aa5adf379abaeb4d43225cd32098a9d7f565fae8206a061042R45-R49